### ISSUE TITLE
Rule Proposal: array-bracket-newline

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6073]
From [validateNewlineAfterArrayElements](http://jscs.info/rule/validateNewlineAfterArrayElements).

This will require/disallow line breaks after open array brackets and before close array brackets.
## Options

``` js
{
    "array-bracket-newline": [
        "error", 
        "always" or "never" or {"multiline": <boolean>, "minItems": <integer>}
    ]
}
```
- `"always"` - Requires line breaks always.
- `"never"` - Disallows line breaks always.
- An object - Requires line breaks if any of properties is satisfied. Otherwise, disallows line breaks.
  - `multiline` (default is `false`) - Requires line breaks if the content is multiline. If this is false, this condition is disabled.
  - `minItems` (default is `0`) - Requires line breaks if the number of elements is more than the given integer. If this is 0, this condition is disabled.
## Examples

``` js
/*eslint array-bracket-newline: ["error", "always"]*/

// ✔ GOOD
let a = [
];
let b = [
    1
];
let c = [
    1, 2
];
let d = [
    1,
    2
];
let e = [
    function foo() {
        dosomething();
    }
];

// ✘ BAD
let a = [];
let b = [1];
let c = [1, 2];
let d = [1, 
    2];
let e = [function foo() {
    dosomething();
}];
```

``` js
/*eslint array-bracket-newline: ["error", "never"]*/

// ✔ GOOD
let a = [];
let b = [1];
let c = [1, 2];
let d = [1,
    2];
let e = [function foo() {
    dosomething();
}];

// ✘ BAD
let a = [
];
let b = [
    1
];
let c = [
    1, 2
];
let d = [
    1, 
    2
];
let e = [
    function foo() {
        dosomething();
    }
];
```

``` js
/*eslint array-bracket-newline: ["error", {"multiline": true}]*/

// ✔ GOOD
let a = [];
let b = [1];
let c = [1, 2];
let d = [
    1, 
    2
];
let e = [
    function foo() {
        dosomething();
    }
];

// ✘ BAD
let a = [
];
let b = [
    1
];
let c = [
    1, 2
];
let d = [1,
    2];
let e = [function foo() {
    dosomething();
}];
```

``` js
/*eslint array-bracket-newline: ["error", {"minItems": 2}]*/

// ✔ GOOD
let a = [];
let b = [1];
let c = [
    1, 2
];
let d = [
    1, 
    2
];
let e = [function foo() {
    dosomething();
}];

// ✘ BAD
let a = [
];
let b = [
    1
];
let c = [1, 2];
let d = [1, 
    2];
let e = [function foo() {
    dosomething();
}];
```

``` js
/*eslint array-bracket-newline: ["error", {"multiline": true, "minItems": 2}]*/

// ✔ GOOD
let a = [];
let b = [1];
let c = [
    1, 2
];
let d = [
    1, 
    2
];
let e = [
    function foo() {
        dosomething();
    }
];

// ✘ BAD
let a = [
];
let b = [
    1
];
let c = [1, 2];
let d = [1,
    2];
let e = [function foo() {
    dosomething();
}];
```
## Related Rules

|  | brace style | space style | line break style |
| --: | :-- | :-- | :-- |
| **block** | [brace-style](http://eslint.org/docs/rules/brace-style) | [block-spacing](http://eslint.org/docs/rules/block-spacing) | [max-statements-per-line](http://eslint.org/docs/rules/max-statements-per-line) |
| **object** | [#6072 object-curly-newline](https://github.com/eslint/eslint/issues/6072) | [object-curly-spacing](http://eslint.org/docs/rules/object-curly-spacing) | [object-property-newline](http://eslint.org/docs/rules/object-property-newline) |
| **array** | [#6073 array-bracket-newline](https://github.com/eslint/eslint/issues/6073) | [array-bracket-spacing](http://eslint.org/docs/rules/array-bracket-spacing) | [#6075 array-element-newline](https://github.com/eslint/eslint/issues/6075) |

**Comment by ilyavolodin** [comment]
If anything, I think this should be merged with #6067 into a single rule.

**Comment by mysticatea** [comment]
Probably you have mistaken the issue number?

I made proposals `newline-in-array-bracket` and `newline-per-array-elements`.
Because we have `object-property-newline` and `lines-per-class-members` (maybe) and line break style rules for each kind of parentheses (maybe).

**Comment by mysticatea** [comment]
Currently, in my head:

| | brace style | space style | line break style |
|:---|:---|:---|:---|
| block | `brace-style` | `block-spacing` | `max-statements-per-line` |
| object | `newline-object-curly` | `object-curly-spacing` | `object-property-newline` |
| array | `newline-array-bracket` | `array-bracket-spacing` | `newline-per-array-elements` (or `array-element-newline`) |

**Comment by ilyavolodin** [comment]
Ah, sorry, you are correct, wrong issue number. I meant combining this with `newline-in-object-curly` proposal.

**Comment by mysticatea** [comment]
I updated this proposal.
Especially, I added the table of related rules. I think that keeping the separation of this and #6072 is fine because the table.

If there are not objection opinions, I want to implement this.

**Comment by mysticatea** [comment]
Ops, `array-bracket-spacing` is covering destructuring's ArrayPattern nodes. So this rule also should cover it as well.
So..., this rule should be able to configure of line breaks each kind (literals / patterns).

```js
{
    "array-bracket-newline": [
        "error", 
        "always" or "never" or {"multiline": <boolean>, "minItems": <integer>}
    ]
    // or
    "array-bracket-newline": [
        "error", 
        {
            "literal": "always" or "never" or {"multiline": <boolean>, "minItems": <integer>},
            "pattern": "always" or "never" or {"multiline": <boolean>, "minItems": <integer>}
        }
    ]
}
```

**Comment by mysticatea** [comment]
@eslint/eslint-team #6072 got accepted, so, how about this rule? This and #6072 are pretty similar.

Thank you.

**Comment by ilyavolodin** [comment]
I'm :+1: but I still wonder if we should merge all of those 3 rules into one.

**Comment by mysticatea** [comment]
Hmm, The implementation of `object-curly-newline` rule and `array-bracket-newline` rule are almost same surely, but I feel merging is an opposite of spacing rule's decision. The 2 spacing rules (`object-curly-spacing` and `array-bracket-spacing`) seemed to be separated from `space-in-brackets`. I'd like to hear reasons for the decision.

**Comment by nzakas** [comment]
Agree with @mysticatea - I think having separate rules is the better approach. 

👍

**Comment by jDmacD** [comment]
Is this still on the cards?

**Comment by platinumazure** [comment]
@jDmacD Yes! And if you're volunteering to write a PR, we will gladly welcome that as well :smile:

**Comment by alberto** [comment]
I'll champion this.

**Comment by kaicataldo** [comment]
@alberto I think you should go for it if you're willing!

**Comment by JPeer264** [comment]
Is this still worth a shot? I can try to implement these features.

**Comment by alberto** [comment]
@JPeer264 sure!

**Comment by JPeer264** [comment]
Nice! Would this option be accepted as well: https://github.com/eslint/eslint/issues/8254 ? This would add `objectsInArrays` and `arraysInArrays`:

```js
{
    "array-bracket-newline": [
        "error", 
        "always" or "never" or {"multiline": <boolean>, "minItems": <integer>, "arraysInArrays": <boolean>, "objectsInObjects": <boolean>}
    ]
}
```

**Comment by not-an-aardvark** [comment]
Could you clarify what that option would do in this context?

**Comment by alberto** [comment]
I don't think so. This is about brace style.
This would be a better fit: https://github.com/eslint/eslint/issues/6075

**Comment by JPeer264** [comment]
@not-an-aardvark @alberto actually, the behavior would already been satisfied with the option `"multiline": true`. I am will give now this feature a shot

**Comment by alberto** [comment]
Ok, maybe I misunderstood what you wanted :)

**Comment by JPeer264** [comment]
I got a bit confused about the base requirements with the option `{ "minItems": 2 }`:

```js
/*eslint array-bracket-newline: ["error", {"minItems": 2}]*/

// ✔ GOOD
let e = [function foo() {
    dosomething();
}];

// ✘ BAD
let e = [function foo() {
    dosomething();
}];
```

Shouldn't `bad` look like following?

```js
// ✘ BAD
let e = [
    function foo() {
        dosomething();
    }
];
```

**Comment by not-an-aardvark** [comment]
This was fixed by https://github.com/eslint/eslint/pull/8314.
